src/java/voldemort/store/readonly/swapper/AdminStoreSwapper.java

```diff
@@ -174,14 +168,52 @@ public String call() throws Exception {
+                private String fetch(String hadoopStoreDirToFetch, AdminClient adminClient, int attempt) {
                     // TODO: Catch specific exception if async task status is not found, and retry in that case.
-                    logger.info("Invoking fetch for " + node.briefToString() + " for " + storeDir);
-                    return adminClient.readonlyOps.fetchStore(node.getId(),
-                            storeName,
-                            storeDir,
-                            pushVersion,
-                            timeoutMs);
+
+                    if (attempt > 1) {
+                        logger.info("Fetch attempt " + attempt + "/" + MAX_FETCH_ATTEMPTS + " for " + node.briefToString()
+                                    + ". Will wait " + WAIT_TIME_BETWEEN_FETCH_ATTEMPTS + " ms before going ahead.");
+                        try {
+                            Thread.sleep(WAIT_TIME_BETWEEN_FETCH_ATTEMPTS );
+                        } catch (InterruptedException e) {
+                            throw new VoldemortException(e);
+                        }
+                    }
+
+                    logger.info("Invoking fetch for " + node.briefToString() + " for " + hadoopStoreDirToFetch);
+                    try {
+                        return adminClient.readonlyOps.fetchStore(node.getId(),
+                                                                  storeName,
+                                                                  hadoopStoreDirToFetch,
+                                                                  pushVersion,
+                                                                  timeoutMs);
+                    } catch (AsyncOperationTimeoutException e) {
+                        throw e;
+                    } catch (VoldemortException ve) {
+                        if (attempt >= MAX_FETCH_ATTEMPTS) {
+                            throw ve;
+                        }
+
+                        if (ExceptionUtils.recursiveClassEquals(ve, ExceptionUtils.BNP_SOFT_ERRORS)) {
+                            String logMessage = "Got a " + ve.getClass().getSimpleName() + " from " + node.briefToString()
+                                                + " while trying to fetch store '" + storeName + "'"
+                                                + " (attempt " + attempt + "/" + MAX_FETCH_ATTEMPTS + ").";
+                            AdminClient possiblyRefreshedAdminClient = adminClient;
+                            if (adminClient.isClusterModified()) {
+                                logMessage += " It seems like the cluster.xml state has changed since this AdminClient"
+                                            + " was constructed. Therefore, we will attempt constructing a fresh"
+                                            + " AdminClient and retrying the fetch operation.";
+                                possiblyRefreshedAdminClient = adminClient.getFreshClient();
```

Comment:
Shall we also need to update this.adminClient to the refreshed one? Looks like we create a new one but and pass it in to recursive "fetch"  but this.adminClient is still the old one.